### DISCUSSION TITLE
[✨] Optimistic Updates

### DISCUSSION THREAD

**Discussion by sacarvy** [QwikDev/qwik-evolution#138] [External Library Ideas]
### Is your feature request related to a problem?

Currently, qwik brings server and client boundaries close!

Yet, is there any guide upon how to implement Optimistic Updates with already loaded data from route loaders,

### Describe the solution you'd like

Hopefully, there is some guidance around this topic in discussion among qwik community

### Describe alternatives you've considered

Alternatively, I created a two separate variables one for loading the data using **loader**, and another just storing the value and updating it, when the data is loaded on client. using Signal

**Comment by shairez** [comment]
Hi @sacarvy and everyone following this discussion! 👋

First off, we want to say thank you for bringing this idea to the table! 

We discussed this proposal in a recent team meeting, and agreed  that this feature will be quite complex to "get right" (some of us tried it in real projects, and it's much more complicated for non trivial cases). We want to make sure every feature we add will sustainable for the long run. 

So instead, we’d love to see this implemented as an external library first. This way, we can see how it performs and get real feedback from the community on its usability and functionality. If it proves successful and maintains stability, we’ll definitely consider adding it as a core feature in the future!

Thank you again for the proposal! Let’s keep the conversation going, and we’re here to support anyone interested in developing this as a library. 😊

@wmertens mind sharing more context about how to get this working as an external library? (thanks)

**Comment by finalgamer** [comment]
We currently use a service worker to catch any `q-data.json` requests that we know the route loader response (an empty response) for already and do not even ask the server for the data.

**Comment by shairez** [comment]
Here's a simple implementation by @mhevery - 
https://github.com/mhevery/qwik-demo/blob/main/src/routes/optimistic/use-optimistic.ts
https://www.loom.com/share/e913fb4fe3304073baec5db74523b2f4

and here's some more context @wmertens shared yesterday:

To create a full solution an algorithm like CRDT needs to be implemented - 
https://en.wikipedia.org/wiki/Conflict-free_replicated_data_type

And here's an implementation of it in JS - 
https://github.com/yjs/yjs

Wout, please feel free to add more context / code examples if you have any
